test/graph_rewrites/test_ng_var_update_ng_tensor.cc

```diff
@@ -0,0 +1,116 @@
+namespace ngraph_bridge {
+namespace testing {
+
+class NGVarUpdateNGTensorOpTest : public tensorflow::OpsTestBase {};
+
+// This test requires the env variable NGRAPH_TF_NGVARIABLE_BUFFER_SHARING=0
+// when running on CPU
+TEST_F(NGVarUpdateNGTensorOpTest, KernelTest) {
+  // Create a normal TF tensor: input_tf_tensor and assign values
+  // This will be used to assign initial value to the TF tensor
+  // that is a part of the NGraph Var resource
+  Tensor input_tf_tensor(DT_FLOAT, TensorShape({2}));
+  auto input_tf_tensor_flat = input_tf_tensor.flat<float>();
+  for (size_t i = 0; i < input_tf_tensor_flat.size(); i++) {
+    input_tf_tensor_flat.data()[i] = 1.0;
+  }
+
+  // Create an NGraphVar resource object
+  NGraphVar* var = new NGraphVar(DT_FLOAT, TensorShape{2}, "CPU");
+  // Assign the ng_tensor with initial value and use it to copy an
+  // initial value to the tf_tensor
+  // After this, both ng_tensor and tf_tensor will hold the same value
+  // i.e. (1.0, 1.0)
+  var->update_ng_tensor(&input_tf_tensor);
+  var->copy_ng_to_tf();
+
+  // Create a normal TF tensor: input_ng_tensor and assign values
+  // This will be used to assign initial value to the NG tensor
+  // that is a part of the NGraph Var resource
+  Tensor input_ng_tensor(DT_FLOAT, TensorShape({2}));
+  auto input_ng_tensor_flat = input_ng_tensor.flat<float>();
+  // Assign a new value to the normal TF tensor: input_tensor
+  for (size_t i = 0; i < input_ng_tensor_flat.size(); i++) {
+    input_ng_tensor_flat.data()[i] = 5.0;
+  }
+
+  // Update the ng_tensor with the new value i.e. (5.0, 5.0)
+  var->update_ng_tensor(&input_ng_tensor);
+  // So now, both ng_tensor and tf_tensor have different values
+  // which is the desired configuration for the test
+
+  // Create NGraphVariableUpdateNGTensor node
+  ASSERT_OK(NodeDefBuilder("sync_node", "NGraphVariableUpdateNGTensor")
+                .Input(FakeInput(DT_FLOAT_REF))
+                .Attr("T", DT_FLOAT)
+                .Attr("ngraph_variable_shared_name", "var1")
+                .Attr("ngraph_graph_id", 1)
+                .Finalize(node_def()));
+  ASSERT_OK(InitOp());
+
+  // Add NGraph resource to the same container as the test op
+  ContainerInfo cinfo_;
+  NodeDef ndef;
+  ndef.set_name("node1");
+  AddNodeAttr("container", "", &ndef);
+  AddNodeAttr("shared_name", "var1", &ndef);
+  ASSERT_OK(cinfo_.Init(device_->resource_manager(), ndef, true));
+
+  ASSERT_OK(device_->resource_manager()->Create<NGraphVar>(cinfo_.container(),
+                                                           cinfo_.name(), var));
+
+  inputs_.push_back({&lock_for_refs_, var->tensor()});
+
+  ASSERT_OK(RunOpKernel());
+
+  shared_ptr<ngraph::runtime::Tensor> ng_t = var->ng_tensor();
+  Tensor output_tensor(DT_FLOAT, TensorShape({2}));
+  void* dst_ptr = DMAHelper::base(&output_tensor);
+  ng_t->read(dst_ptr, 0, output_tensor.TotalBytes());
+
+  ASSERT_EQ(output_tensor.DebugString(), input_tf_tensor.DebugString());
+
```

Comment:
Is there a way to get the output of the sync op, may be we can compare that too?